dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/jdbc/JdbcProgramOrgUnitAssociationsStore.java

```diff
@@ -0,0 +1,231 @@
+        "    pr.sharing, " +
+        "    array_agg(ou.uid) agg_ou_uid " +
+        "from program pr " +
+        "    left outer join program_organisationunits po on pr.programid = po.programid " +
+        "    left outer join organisationunit ou on po.organisationunitid = ou.organisationunitid " +
+        "where";
+
+    private static final String INNER_QUERY_GROUPING_BY = "group by pr.uid, pr.sharing";
+
+    private final CurrentUserService currentUserService;
+
+    private final JdbcTemplate jdbcTemplate;
+
+    public IdentifiableObjectAssociations getProgramOrganisationUnitsAssociations( Set<String> programUids )
+    {
+
+        Set<String> userOrgUnitPaths = getUserOrgUnitPaths();
+
+        return jdbcTemplate.query(
+            buildSqlQuery( programUids, userOrgUnitPaths, currentUserService.getCurrentUser() ),
+            resultSet -> {
+                IdentifiableObjectAssociations identifiableObjectAssociations = new IdentifiableObjectAssociations();
+                while ( resultSet.next() )
+                {
+                    identifiableObjectAssociations.addAllAssociations(
+                        resultSet.getString( 1 ),
+                        Arrays.asList( (String[]) resultSet.getArray( 2 ).getArray() ) );
+
+                }
+                return identifiableObjectAssociations;
+            } );
+    }
+
+    private Set<String> getUserOrgUnitPaths()
+    {
+        return currentUserService.getCurrentUserOrganisationUnits().stream()
+            .map( OrganisationUnit::getPath )
+            .collect( Collectors.toSet() );
+    }
+
+    private String buildSqlQuery( Set<String> programUids, Set<String> userOrgUnitPaths, User currentUser )
+    {
+        Stream<String> queryParts = Stream.of(
+            SHARING_OUTER_QUERY_BEGIN,
+            innerQueryProvider( programUids, userOrgUnitPaths, currentUser ),
+            SHARING_OUTER_QUERY_END );
+
+        if ( nonSuperUser( currentUser ) )
+        {
+            queryParts = Stream.concat(
+                queryParts,
+                Stream.of(
+                    "where",
+                    getSharingConditions( LIKE_READ_METADATA ) ) );
+        }
+        return queryParts.collect( joining( " " ) );
+    }
+
+    private String innerQueryProvider( Set<String> programUids, Set<String> userOrgUnitPaths, User currentUser )
+    {
+        Stream<String> queryParts = Stream.of(
+            INNER_SQL_QUERY,
+            getProgramUidsFilter( programUids ) );
+
+        if ( nonSuperUser( currentUser ) )
+        {
+            queryParts = Stream.concat( queryParts,
+                Stream.of(
+                    "and",
+                    getUserOrgUnitPathsFilter( userOrgUnitPaths, currentUser ) ) );
+        }
+
+        queryParts = Stream.concat( queryParts, Stream.of( INNER_QUERY_GROUPING_BY ) );
+
+        return queryParts.collect( joining( " " ) );
+    }
+
+    private String getSharingConditions( String access )
```

Comment:
It would make sense to have all this sharing related methods in a Utility class?

Reply:
it makes absolutely sense, I'll move all query building related code in a separate class